src/Dialect/ONNX/Rewrite.td

```diff
@@ -390,6 +394,77 @@ def FuseMulConvNullBiasPattern: Pat<
+                        (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_),
+                        $shape))),
+  (ONNXExpandOp (ONNXAddOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+
+// Pattern to turn `Mul of Constant and Expand` into `Expand of Mul of Constants`
+// which put constants together so that Mul can be folded.
+// Note: look like tablegen failed with `error: encountered non-op, non-NativeCodeCall match.`
+// when having the second pattern using `either`.
+// As a workaround, we use two patterns.
+def PropagateScalarConstantExpandForMul1: Pat<
+  (ONNXMulOp (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                     (ONNXExpandOp
+                        (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_),
+                        $shape)),
+  (ONNXExpandOp (ONNXMulOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+def PropagateScalarConstantExpandForMul2: Pat<
+  (ONNXMulOp (ONNXExpandOp
+                (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                $shape),
+             (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_)),
+  (ONNXExpandOp (ONNXMulOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+
+// Pattern to turn `Sub of Constant and Expand` into `Expand of Sub of Constants`
+// which put constants together so that Sub can be folded.
+def PropagateScalarConstantExpandForSub1: Pat<
+  (ONNXSubOp (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                     (ONNXExpandOp
+                        (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_),
+                        $shape)),
+  (ONNXExpandOp (ONNXSubOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+def PropagateScalarConstantExpandForSub2: Pat<
+  (ONNXSubOp (ONNXExpandOp
+                (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                $shape),
+             (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_)),
+  (ONNXExpandOp (ONNXSubOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+
+// Pattern to turn `Div of Constant and Expand` into `Expand of Div of Constants`
+// which put constants together so that Div can be folded.
+def PropagateScalarConstantExpandForDiv1: Pat<
+  (ONNXDivOp (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                     (ONNXExpandOp
+                        (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_),
+                        $shape)),
+  (ONNXExpandOp (ONNXDivOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+def PropagateScalarConstantExpandForDiv2: Pat<
+  (ONNXDivOp (ONNXExpandOp
+                (ONNXConstantOp:$lhs $_, $_, $_, $_, $_, $_, $_, $_),
+                $shape),
+             (ONNXConstantOp:$rhs $_, $_, $_, $_, $_, $_, $_, $_)),
+  (ONNXExpandOp (ONNXDivOp $lhs, $rhs), $shape),
+  [(IsScalarTensor:$lhs), (IsScalarTensor:$rhs)]
+>;
+
```

Comment:
If you agree that this is true, maybe add a comment that "this pattern can be reproduce for all element wise operations".

@sorenlassen @tungld I see that you require both constants to be scalar, is that a necessary requirement?

Reply:
in, e.g., Sub(C1,Expand(C2,shape)), if C1 is not scalar then you don't know how it will play together with shape - for instance, it could be that shape broadcasts to C1's larger shape and not the other way around

it's probably possible to define more lax conditions but it makes things more complicated

it might be safe to do the transform to Expand(Sub(C1,C2),shape) if C2 is not scalar

@tungld I think you need to restrict isScalarTensor to requite rank=0 (or find another workaround), otherwise the Sub example breaks if C1 has rank 1 and C2 and shape have rank 0

fwiw, if you want to instantiate this pattern for more ops without code duplication, you could use a class like FuseTransposeAndElementwiseUnaryOpPattern in the old Rewrite.td

Reply:
> otherwise the Sub example breaks if C1 has rank 1 and C2 and shape have rank 0

I have a lit test in this PR for this case, it works when applying constprop, e.g.

Input:
```mlir
func.func @test_replace_sub_by_expand_2(%arg0: tensor<?xi64>) -> tensor<2x?xf32> {
    %0 = onnx.Constant dense<2.000000e+00> : tensor<f32>
    %1 = onnx.Constant dense<5.000000e+00> : tensor<1xf32>
    %2 = onnx.Constant dense<2> : tensor<1xi64>
    %3 = "onnx.Dim"(%arg0) {axis = 0 : si64} : (tensor<?xi64>) -> tensor<1xi64>
    %4 = "onnx.Concat"(%2, %3) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
    %5 = "onnx.Sub"(%1, %0) : (tensor<1xf32>, tensor<f32>) -> tensor<1xf32>
    %6 = "onnx.Expand"(%5, %4) : (tensor<1xf32>, tensor<2xi64>) -> tensor<2x?xf32>
    return %6 : tensor<2x?xf32>
  }
```

output when doing constprop:
```mlir
func.func @test_replace_sub_by_expand_2(%arg0: tensor<?xi64>) -> tensor<2x?xf32> {
    %0 = onnx.Constant dense<3.000000e+00> : tensor<1xf32>
    %1 = onnx.Constant dense<2> : tensor<1xi64>
    %2 = "onnx.Dim"(%arg0) {axis = 0 : si64} : (tensor<?xi64>) -> tensor<1xi64>
    %3 = "onnx.Concat"(%1, %2) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
    %4 = "onnx.Expand"(%0, %3) : (tensor<1xf32>, tensor<2xi64>) -> tensor<2x?xf32>
    return %4 : tensor<2x?xf32>
  }
```

Reply:
I can generalize it to a common pattern.

Reply:
the example I had in mind was
```
    %1 = onnx.Constant dense<1> : tensor<1xi32>
    %2 = onnx.Constant dense<2> : tensor<i32>
    %3 = onnx.Constant dense<> : tensor<0xi64>
    %4 = "onnx.Expand"(%2, %3) : (tensor<i32>, tensor<0xi64>) -> tensor<i32>
    %5 = "onnx.Sub"(%1, %4) : (tensor<1xi32>, tensor<i32>) -> tensor<1xi32>
```
it goes wrong if the PropagateScalarConstantExpandForSub1 pattern rewrites that to
```
    %1 = onnx.Constant dense<1> : tensor<1xi32>
    %2 = onnx.Constant dense<2> : tensor<i32>
    %3 = onnx.Constant dense<> : tensor<0xi64>
    %5 = "onnx.Sub"(%1, %2) : (tensor<1xi32>, tensor<i32>) -> tensor<1xi32>
    %6 = "onnx.Expand"(%5, %3) : (tensor<1xi32>, tensor<0xi64>) -> tensor<i32>
```
because the empty pattern `%3` cannot expand the shape `[1]`

Reply:
What is the semantics of expand when shape is `tensor<0xi64`?

When I run your example with this patch, the output is
```
func.func @test_replace_sub_by_expand_rank0() -> tensor<1xi32> {
    %0 = onnx.Constant dense<1> : tensor<1xi32>
    %1 = onnx.Constant dense<2> : tensor<i32>
    %2 = onnx.Constant dense<> : tensor<0xi64>
    %3 = "onnx.Sub"(%0, %1) : (tensor<1xi32>, tensor<i32>) -> tensor<1xi32>
    %4 = "onnx.Expand"(%3, %2) : (tensor<1xi32>, tensor<0xi64>) -> tensor<1xi32>
    return %4 : tensor<1xi32>
  }
```

Note that the last op is the output type of `%4` is `tensor<1xi32` not `tensor<i32` as your output. I am not sure how you know the output type of `%6` is `tensor<i32` (`%6 = "onnx.Expand"(%5, %3) : (tensor<1xi32>, tensor<0xi64>) -> tensor<i32>`)

Reply:
@AlexandreEichenberger @sorenlassen I created a C++ rewriting rule for binary ops instead of using tablegen. It can be extended to support new binary ops in the future.

Reply:
> What is the semantics of expand when shape is `tensor<0xi64`?

I haven't tested it but I assume that you can only expand the empty shape to the empty shape, i.e., Expand(X,shape=[]) only works if X has a rank 0 type

> I am not sure how you know the output type of `%6` is `tensor<i32 (%6 = "onnx.Expand"(%5, %3) : (tensor<1xi32>, tensor<0xi64>) -> tensor<i32>)`

sorry for being unclear, I should have omitted `-> tensor<i32>`, the point of the example is that the last line is breaks down because Expand(X,shape=[]) doesn't work when X has a rank 1 type

Reply:
OK, to avoid confusion, I exclude the case where Expand's shape is `tensor<0xdtype>`.

Reply:
sorry, please forget my example, I misunderstood how Expand works, I thought it's shape argument was the shape of the output, but on closer inspection I found that it broadcasts together with the input shape to form the output shape

in PR #2508 I added an Expand const prop lit test to illustrate this